### ISSUE TITLE
On Play 2.8.18, heap memory issues arises on server

### ISSUE THREAD

**Issue by atilparikh** [playframework/playframework#11526]
I was using play in my production server (more than 2 years) with below configuration
```
-J-Xms1023m 
-J-Xmx2048m 
-J-javaagent:newrelic/newrelic.jar 
-Dnewrelic.environment=aptServer 
-J-XX:+UseStringDeduplication 
-J-server 
-Dhttp.port=9879
```

And it was working fine, but after updating it with 2.8.18; in 24hrs only it is crashed and asking heap memory issues; I have changed Xmx to 4G also; even though it is crashing. Doing some more optimisation, changes GCs but nothing helps, tried to remove newrelic also but nothing works.

When I port it back to 2.8.17, it works like a gem. Some dependency updation in Play 2.8.18 may be a problem point or something.

### Play Version
Issues in Play 2.8.18
Sbt 1.7.2
Scala 2.13.10

### API
Using PlayJava using Java 17

### Operating System
Ubuntu 20.04.5 LTS
Linux ip-172-31-43-36 5.13.0-1031-aws #35~20.04.1-Ubuntu SMP Mon Jun 13 22:30:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### JDK
java version "17.0.1" 2021-10-19 LTS
Java(TM) SE Runtime Environment (build 17.0.1+12-LTS-39)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.1+12-LTS-39, mixed mode, sharing)

### Expected Behavior
Memory handling should be optimal enough to release unused memory and some intervals

**Comment by mkurz** [comment]
Can you please post the exception that's occuring?

So here is the diff from 2.8.17 to 2.8.18: https://github.com/playframework/playframework/compare/2.8.17...2.8.18

- Which server backend do you use? Netty or akka-http?
- Does you application handle multipart/form-data?

**Comment by atilparikh** [comment]
In log, I was getting...

2022-11-06 21:43:29 +0000 [ERROR] from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-14 - exception on LARS’ timer thread
java.lang.OutOfMemoryError: Java heap space
2022-11-06 21:43:29 +0000 [INFO] from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-14 - starting new LARS thread
2022-11-06 21:43:29 +0000 [ERROR] from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-10 - Uncaught error from thread [application-akka.actor.internal-dispatcher-4]: Java heap space, shutting down ActorSystem[application]
java.lang.OutOfMemoryError: Java heap space
2022-11-06 21:43:29 +0000 [ERROR] from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-10 - Uncaught error from thread [application-scheduler-1]: Java heap space, shutting down ActorSystem[application]
java.lang.OutOfMemoryError: Java heap space
2022-11-06 21:43:29 +0000 [ERROR] from akka.actor.ActorSystemImpl in application-akka.actor.default-dispatcher-10 - Uncaught error from thread [application-akka.actor.default-dispatcher-12]: Java heap space, shutting down ActorSystem[application]
java.lang.OutOfMemoryError: Java heap space
2022-11-06 21:43:29 +0000 [INFO] from akka.actor.CoordinatedShutdown in application-akka.actor.default-dispatcher-12 - Running CoordinatedShutdown with reason [ActorSystemTerminateReason]
2022-11-06 21:43:29 +0000 [INFO] from play.core.server.AkkaHttpServer in application-scheduler-1 - Stopping Akka HTTP server...

We are using **AkkaHttp** server and yes our application is **using multipart/form-data** also.

**Comment by mkurz** [comment]
I see you use newrelic, does that give some information what is using all the memory?

**Comment by atilparikh** [comment]
Didn't find much through newrelic. Few logs I found are almost same like above. This system is more over campaigning system for sending emails and checking their opening and all.

I found few stuff...
HikariPool-3 - Thread starvation or clock leap detected (housekeeper delta=3m557ms568 s256ns).
HikariPool-2 - Thread starvation or clock leap detected (housekeeper delta=3m45772ms630 s72ns).
Internal server error, sending 500 response
starting new LARS thread
exception on LARS' timer thread

**Comment by mkurz** [comment]
And you did only change the Play version, nothing else? Any other code changes, any changes to jvm flags or sbt configs/flags? Java version change?

**Comment by mkurz** [comment]
Also regarding `Internal server error, sending 500 response`, would be nice if you could log that error.

**Comment by mkurz** [comment]
Not sure if HikariCP `Thread starvation` even matters, see discussion:
- https://github.com/brettwooldridge/HikariCP/issues/679
- https://stackoverflow.com/questions/38703876/log-warning-thread-starvation-or-clock-leap-detected-housekeeper-delta-springh

**Comment by jtjeferreira** [comment]
@atilparikh, any chance you can add the JVM flag `-XX:+HeapDumpOnOutOfMemoryError` and then collect the heap dump when the application crashes?

**Comment by atilparikh** [comment]
@mkurz yes there is no other changes than this, and for HickariCP I have checked earlier also, it was not the issue, checked again yesterday, but none.

@jtjeferreira I will integrate it in a day with production; and let's hope we shall have something concrete to work on.

**Comment by atilparikh** [comment]
<img width="657" alt="Screenshot 2022-11-08 at 16 09 29" src="https://user-images.githubusercontent.com/70359650/200744420-94f4645c-ab19-4df7-a277-1da3e3dde35f.png">
<img width="657" alt="Screenshot 2022-11-08 at 16 09 08" src="https://user-images.githubusercontent.com/70359650/200744423-c060c356-1f5b-4849-9130-c31eb4ae6fdd.png">

This is what happened yesterday, I have taken screenshot from NewRelic, but due to some requirement, server re-deployed.  So couldn't reproduce out of memory heap; but seems this is the starting point of that.  I have check that trigger time -60m to +10m and found that server was almost idle.

**Comment by mkurz** [comment]
It's hard to tell what's going on. Waiting for the OOM report.
Meanwhile, can you see in NewRelic which classes survive garbage collection?

Also, what happens around 11:35 ? Increase of traffic? How many requests is this app handling? You say it's for email, so is it just serving 1pixel images to check if an email was opened?

**Comment by pk044** [comment]
@atilparikh have you tried testing it on ParallelGC? I had similar problems with running G1GC with Play - memory didn't seem to clear up, and it went OOM after a day/few days. Switching to ParallelGC fixed that.

**Comment by atilparikh** [comment]
I agree @mkurz without OOM report it is really hard to find who is the culprit. And I have checked at that time with no unusual data. And yes, this server is responsible to sending emails, serving 1 pixel image and few Akka jobs to update some info inside RDBMS.

@pk044 I have tried ZGC with 2.8.18, but not Parallel one. I shall try it also, and send you findings too.

**Comment by mkurz** [comment]
@atilparikh So where does you app handle `multipart/form-data` then if it's only serving 1 pixel images? We did some changes to to how to parse multipart/form-data payloads in 2.8.18, but if you are not using that I am a bit lost what's going on.
Let's wait for the crash report, maybe we can find out more.

**Comment by atilparikh** [comment]
Our internal team is using multipart/form-data to upload few data related to campaigns, like campaign details and other stuff like target, criteria etc. Yes, I am also waiting for the same.

**Comment by mkurz** [comment]
OK, but how often do they upload anything? Like once per day? I doubt that is the cause of the problem.

**Comment by atilparikh** [comment]
Yes maximum 2-3 times a day.

**Comment by mkurz** [comment]
> Yes maximum 2-3 times a day.

OK, I somehow doubt that will be the problem. Do you know what happened at around 1135 in the screenshot you posted? Is it just that around midday people open their newsletters?

**Comment by atilparikh** [comment]
@mkurz this system is continuous, we have global audience, so traffic is almost same at every time bit up and down depends on demography density.  And checked from 1100hrs to 1145hrs, nothing very special / different I found.

**Comment by atilparikh** [comment]
Found yesterday's dump, when I was trying with zgc. 
<img width="1400" alt="Screenshot 2022-11-11 at 13 23 53" src="https://user-images.githubusercontent.com/70359650/201309080-66bf6277-3cc6-4b2f-b66c-a5f80fe51315.png">

**Comment by lenin122** [comment]
We have simmilar issuu, play 2.8.18, newRelic 7.11.0
When server starts, it blows to top of heap in 2 minutes. The most strange thing, is that it was working on production almowst a week without any issues. And now without any extra load we see that G1 Old space is filled same as heap ~5Gb, no 00M issue or outOfMemory, server is restarted by healthchecker. But, in new relic akka-dispatcher started to show increased amount of time usage.
![image](https://user-images.githubusercontent.com/8101083/204629905-15d6fdcd-4198-4fde-979e-322d0536b397.png)
Release of our app was 1 week before this started

**Comment by atilparikh** [comment]
@lenin122 yes exactly same stuff, not able to find exact cause, but few changes I have done now things better but not fully resolved

Changed Xmx to 9G from 2G
Set MinHeapFreeRatio to 20
Set MaxHeapFreeRatio to 40
And ZGC literally sucks, so still continue with G1GC.

It may help to resolve

**Comment by mkurz** [comment]
Thanks @lenin122. If you find out more please keep us postet.

**Comment by lenin122** [comment]
> Thanks @lenin122. If you find out more please keep us postet.

Yeap, we found issue in our code. So for us play update did  not produce any problem yet

**Comment by mkurz** [comment]
> So for us play update did not produce any problem yet

Good to hear, thanks :+1:

**Comment by atilparikh** [comment]
@mkurz @lenin122 after lot many permutation and combination, I found that this issues now arises with older play framework also; secondly, lot many parellal serves we have started with different GC configuration/Play version, but we didn't succeed, heavy optimisation changes code wise is happened over a month and improves a lot but problem not resolved. 

Considering above stuff, I am thinking this should not be blocking for next milestone for play. @mkurz it is over to you to close this issue.

Thanks for all of your support @mkurz @lenin122 @jtjeferreira @pk044 and everyone.
